### ISSUE TITLE
Would "mobile outline" be good?

### ISSUE THREAD

**Issue by newjie** [gantry/gantry5#1341]
On many occasions I'd like the a same webpages to have quite different layout/section/module/particles for mobile and desktop view. Currently I don't see any ways to do this other than using media query classes to hide some section on mobile devices. But there would actually make users download unnecessary data that won't eventually display, especially when the section to be hidden on mobile view has big images or video players. So I am thinking it will be perfect if each outline can have a mobile version, I think there should be many solid PHP ways to determine whether users are using mobile devices or not before gantry 5 decides which version to output.

**Comment by adi8i** [comment]
That means you have to add mobile detect in gantry core, then a new set of options in admin, to chose where a particle will be displayed. Is not a bad idea and i don't think it's too complex, am i right? 

You are going to finish inheritance soon, I told you !!!

**Comment by w00fz** [comment]
I think that should be totally doable, one outline for each mobile/desktop and like adi8 said, mobile detect implementation.

The only complicated portion is telling Gantry how to load one outline instead of another based on the mobile detection (desktop vs mobile).

**Comment by adi8i** [comment]
Nah, I think outlines will go nuts, think about WordPress, where you have to create 5,6 outlines for a bigger website. 

I was think to add: 

```
{%- if gantry.mobile_detect.isMobile() and particle.render.mobile -%}
    Render the particle 
{%- endif -%}
```

While thinking of it, i found other issues, and doesn't look doable because if there's one particle per row you have to exclude the whole `<g-grid>`, if there are multiple particle, you have to exclude `<g-block>`

**Comment by mahagr** [comment]
It is doable, but might need some changes in Gantry plugins in both Joomla and WP.

I changed my mind -- it would be great to at least have support for changing outline from outside Gantry, so I will make this a feature request even if we end up implementing/documenting only the part that allows people to do it.

**Comment by sural98** [comment]
Your idea is based on a 1-language Joomla installation. We will need to create as many outlines as the number of languages if we have a multi lan

**Comment by gyvermac** [comment]
You don't need to create as many outlines as number of languages in a multi language installation if yo don't use "language specific" particles. All data coming from the CMS (Joomla-WP-GRAV) are already "linked" to a language.
Because of that in a Joomla multilanguage web site right now it's easier to use "modules" than "particles"...
In this [2 languages web site](http://wild-geese-aviation.com) (under work) I have only 1 outline for the home page and 1 for the others.

**Comment by JoomFX** [comment]
@gyvermac 

> Because of that in a Joomla multilanguage web site right now it's easier to use "modules" than "particles"...

I guess you mean to use the "Gantry 5 Particle" module in the Joomla Module Manager, rather than adding the particle directly in the Layout Manager :wink: 

In my opinion, the best way to configure a Joomla website using G5 is:

1. Create your Module Positions in the Layout Manager
2. Use the "Gantry 5 Particle" module in the Module Manager to publish a particle

Using the above approach you really do NOT need to create a new Outline for each language.

I encourage everybody to use particles instead of modules :smile:

**Comment by adi8i** [comment]
In joomla you don't need that many outlines even if it's a multilanguage website, here the real problem is WP. 

An alternative could be what I mentioned earlier, by adding 3 checkboxes, `render on mobile`, `render on tablet`, `render on desktop`, this way you can chose on which devices a particle will render.

Both solutions have downsides.

**Comment by gyvermac** [comment]
Yes the "Gantry 5 Particle" is a way to "link" some G5 particles to a language when they are language specific. May be better sometimes than cloning a complet outline.

**Comment by mahagr** [comment]
Right now binding particles to a language only works in Joomla (as a module).

**Comment by JoomFX** [comment]
@adi8i agree :wink: 

@gyvermac I'm speaking in general, not only for Multilingual Joomla websites.
Since the native Joomla way is to assign a module to a menu item, then (in my opinion) the most logical and intuitive approach is to create your **Layout** in the Layout Manager (create your module positions) and then use the Joomla's Module Manager and the "Gantry 5 Particle" module to build your website.
Speaking for Joomla only, I think that a website should be build by using the "Gantry 5 Particle" module, rather than dropping the particle directly in the Layout Manager. Once again (in my opinion), in the Layout Manager you should have only your module positions and the static content, which is the "Menu" particle, the "Logo" particle, the "Mobile Menu" particle and probably couple others.

Of course, everybody can build their website by using the approach they like the most. There is nothing wrong to drop the particle directly in the Layout Manager. If you use this approach, you'll need to have more Outlines. I even think that this was the original idea behind G5's Layout Manager - to build your pages there. But it is not that easy to escape from the native CMS logic and to re-educate millions of users who already know that you assign a module (particle) to a menu item.

**Comment by gyvermac** [comment]
@JoomFX I agree concerning Joomla because of the CMS logic. That way you also benefit of functionalities like ACL, languages... But it seems that most templates doesn't encourage to use this approach even when they are in a full package with joomla + demo data.

**Comment by JoomFX** [comment]
@gyvermac Exactly (ACL and other stuff coming from the CMS) :wink: 

Regarding the templates - all our templates and all RT template are built using this approach (with the "Gantry 5 Particle" module, rather than dropping the particle directly in the Layout Manager, except the static particles).

**Comment by gyvermac** [comment]
May be I should subscribe to your Club  :-) But it's out of topic here !

**Comment by w00fz** [comment]
Renamed the title based on https://github.com/gantry/gantry5/issues/1341#issuecomment-224501087

@newjie, reading again this issue and after a month has passed, I personally think that your suggestion is not bad but goes against Gantry philosophy where we promote responsive over mobile-first/mobile-only.

This is not to say that mobile-first/mobile-only is bad, in fact, it would be nice if we could add support for it and probably not even that complicated. I can see the implementation made in the form you suggested, where instead of just a Layout, we now have multiple Layouts based on breakpoints (1 for desktop, 1 for tablets, 1 for mobile, etc).
What needs to be understood is that this approach would bring some limitations and troubles. Being now mobile-first, once a browser loads the page based upon its viewport, nothing else can change, even after the viewport gets resized. This is the opposite of responsive where, like it is now, if you load Desktop and resize down, the layout adapts. In mobile first there is no such thing as adapting the way up, what you load first is what you get always.

Imagine you have in your Tablet Layout a grid with 2 logos and in your Desktop Layout you have a grid with 3 logos. You are on your computer running Chrome and the browser size is matching a tablet-size viewport. Even if you now try to resize your browser bigger, your logos in that grid will always be 2, because that's what it has loaded with. The only way to get that extra 3rd logo is by resizing bigger your browser and reload the page. You can see how this is very different from a responsive layout where elements show up and hide based on media queries.

Entering the mobile-first territory while maintaining responsive layouts can definitely benefit some aspects, because mobile-first is definitely faster than responsive and my example above could be something that a developer might actually want, but will most definitely also cause some troubles. 

Not to mention the huge amount of extra work that will have to go into creating templates. Now a theme developer will have to style for Desktop + Responsive, Tablet + Responsive, Mobile + Responsive.

This is something that should be discussed thoroughly and not rashly implemented or taken lightly.

This is my point of view on the matters and I hope I made it clear where I stand at. 😃

**Comment by joomlabeat** [comment]
Any update on this? I think device detection to allow loading of a different layout is essential - and actually should not be confused with the responsive design/mobile-first approach. Both responsive/mobile-first are on the design side of things and are happening on the client side - they are still useful even in the case of device detection. 

Device detection happens on the server side where php will discover what is the device/browser of the user and will use defined logic about what layouts/elements should render. Still those layouts will obey to any design responsive rules (either it will be mobile first or just responsive from the opposite). 

For example in my case - I have in main 3 columns - component and 2 columns for modules. I want in tablet to **remove** the 1 column and in mobile both 2 columns. I highlight the word **remove** and not saying **hide** as it is a different thing. This is essentially loading a different layout for these 2 different devices types that won't include those extra columns. This is happening once on the server side.

A mobile device for the full time that it will be browsing the website will never become a desktop device, so there is no matter of later needing to display a different scaled up layout for a bigger device. 

For gantry - that could be an option to generate a tablet / mobile version for a layout if needed or the suggestion that I read somewhere in this issue - with checkboxes on sections / grid level.

**Comment by mahagr** [comment]
I still don't think that Gantry should be detecting the browser/device should be done by Gantry -- the issue is not only the layout but also all the modules which are being loaded. Gantry does what Joomla asks it to do, making it out of context for Gantry. 

I bet that Joomla has a few extensions which allow template style and modules to be customized by a device. If those work in the Joomla way, they should work with Gantry out of the box.

**Comment by joomlabeat** [comment]
@mahagr :: I finally tend to agree with you after implementing this in the way you suggested above.

I thought I posted here that I found my way around this, but today I don't see this post.
So with a system plugin that makes use of the [mobile_detect library](http://mobiledetect.net) - although Joomla's browser detection ability can be used as well, I do device detection early on the application execution flow, and then set another gantry template style for tablets and mobiles with `JFactory::getApplication()->setTemplate($templateName, $params);`

For this to work with gantry5 template styles, it requires to pass in the $params the gantry the template style id for the configuration property. Something like `$params = array('configuration'=>STYLE_ID);`.

**Comment by mahagr** [comment]
Yes, the function call needs the second parameter to contain params from a template style. You may want to load template style from the table manually and use the params from there.
